knitr/curse-dims/curse-dims.Rmd

```diff
@@ -0,0 +1,770 @@
+  geom_point() +
+  ylab("Pr[y in hyperball | y in hypercube]") +
+  xlab("number of dimensions") +
+  ggtitle("The Mass is in the Corners");
+plot_corners;
+```
+
+
+## 3. The Normal Distribution
+
+The normal log density for a variate $y \in \mathbb{R}$ with location
+$\mu \in \mathbb{R}$ and scale $\sigma \in \mathbb{R}^+$ is defined by
+
+\[
+\mathsf{Normal}(y \, | \, \mu, \sigma)
+= \frac{1}{\sqrt{2 \pi}}
+  \, \frac{1}{\sigma} 
+  \, \exp
+       \left( 
+         -\frac{1}{2} 
+         \, \left( \frac{y - \mu}{\sigma} \right)^2
+       \right).
+\]
+
+For the unit normal, $\mathsf{Normal}(0, 1)$, many terms drop out, and
+we are left with
+
+\[
+\mathsf{Normal}(y \, | \, 0, 1)
+= \frac{1}{\sqrt{2 \pi}}
+  \, \exp \left( -\frac{1}{2} \ y^2 \right).
+\]
+
+Converting to the log scale, we have
+
+\[
+\log \mathsf{Normal}(y \, | \, 0, 1)
+\ = \ -\frac{1}{2} \ y^2 + \mathrm{constant}
+\]
+
+where the constant does not depend on $y$.  In this form, it's easy to
+see the relation between the unit normal and distance.
+
+
+
+## 4. Vectors of Random Unit Normals
+
+We are now going to generate a random vector $y = (y_1, \ldots, y_D)
+\in \mathbb{R}^D$ by generating each dimension independently as
+
+\[
+y_d \sim \mathsf{Normal}(0, 1).
+\]
+
+The density over vectors is then defined by
+
+\[
+p(y) 
+\ = \
+\prod_{d=1}^D p(y_d)
+\ = \
+\prod_{d=1}^D \mathsf{Normal}(y_d \, | \, 0, 1).
+\]
+
+and on the log scale, that's just  is just
+
+\[
+\log p(y)
+\ = \
+\sum_{d=1}^D -\frac{1}{2} \, y_d^2 + \mathrm{const}
+\ = \
+-\frac{1}{2} \, {|| y ||}^2 + \mathrm{const}.
+\]
+
+Equivalently, we could generate the vector $y$ all at once from a
+multivariate normal with unit covariance matrix,
+
+\[
+y \sim \mathsf{MultiNormal}(\mathbf{0}, \mathbf{I}),
+\]
+
+where $\mathbf{0}$ is a $D$-vector of zero values, and $\mathbf{I}$ is
+the $D \times D$ unit covariance matrix with unit values on the diagonal
+and zero values off diagonal (i.e., unit scale and no correlation).
+
+To generate a random vector $y$ in R, we can use the `rnorm()`
+function, which generates univarite random draws from a normal
+distribution.
+
+```{r comment=NA}
+D <- 10;
+u <- rnorm(D, 0, 1);
+print(u, digits=2);
+```
+
+It is equally straightforward to compute the Euclidean length of `u`:
+
+```{r comment=NA}
+print(euclidean_length(u), digits=3);
+```
+
+What is the distribution of the lengths of vectors generated this way
+as a function of the dimensionality?  We answer the question
+analytically in the next section, but for now, we'll get a handle on
+what's going on through simulation as the dimensionality $D$ grows.
+
+```{r comment=NA}
+log_sum_exp <- function(u) max(u) + log(sum(exp(u - max(u))));
+N <- 1e4;
+D <- 17;
+dim <- rep(NA, D);
+lower <- rep(NA, D);
+middle <- rep(NA, D);
+upper <- rep(NA, D);
+lower_ll <- rep(NA, D);
+middle_ll <- rep(NA, D);
+upper_ll <- rep(NA, D);
+
+mean_ll <- rep(NA, D);
+max_ll <- rep(NA, D);
+for (k in 1:D) {
+  d <- as.integer(sqrt(2)^(k - 1));
```

Comment:
It's the same problem as the autodiff paper.  That operation
isn't dense in the progression I have multiplying by sqrt(2)
rather than 2 to keep everything even on the log2 scale.

I'll just hardcode the sequence and sizes here as
it'll be less confusing.

On Jun 16, 2016, at 6:14 PM, Daniel Lee <notifications@github.com> wrote:
> 
> In knitr/curse-dims/curse-dims.Rmd:
> 
> > +```{r comment=NA}
> > +log_sum_exp <- function(u) max(u) + log(sum(exp(u - max(u))));
> > +N <- 1e4;
> > +D <- 17;
> > +dim <- rep(NA, D);
> > +lower <- rep(NA, D);
> > +middle <- rep(NA, D);
> > +upper <- rep(NA, D);
> > +lower_ll <- rep(NA, D);
> > +middle_ll <- rep(NA, D);
> > +upper_ll <- rep(NA, D);
> > +
> > +mean_ll <- rep(NA, D);
> > +max_ll <- rep(NA, D);
> > +for (k in 1:D) {
> > +  d <- as.integer(sqrt(2)^(k - 1));
> 
> I don't understand this. For k = 1:D, it produces:
> 1, 1, 2, 2, 4, 5, 8, 11, 16, 22, 32, 45, 64, 90, 128, 181, 256.
> 
> The real question is why this is repeating 1 and 2 twice?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>